File: mcp-gateway/events.py
"""Event logger for MCP Gateway — writes NDJSON to MCP_EVENTS_LOG_PATH.

One line per RPC: method, tool, identity, team, server, latency, status, action, etc.
Does not log raw params/results by default (mcp_log_payloads=false).
"""
import json
import logging
import os
import uuid
from datetime import datetime, timezone
from typing import Any, Optional

logger = logging.getLogger(__name__)

_log_path = None
_log_payloads = False


def _get_log_path() -> str:
    global _log_path
    if _log_path is not None:
        return _log_path
    _log_path = os.environ.get("MCP_EVENTS_LOG_PATH", "data/mcp/events.ndjson")
    _log_payloads = os.environ.get("MCP_LOG_PAYLOADS", "false").lower() in ("true", "1", "yes")
    return _log_path


def _is_log_payloads() -> bool:
    global _log_payloads
    if _log_path is None:
        _get_log_path()
    return _log_payloads


def generate_request_id() -> str:
    """Generate a unique request ID."""
    return uuid.uuid4().hex[:16]


def _lookup_name(collection: str, entity_id: Optional[int]) -> Optional[str]:
    """Look up an entity name by ID from the config bundle.

    Best-effort: returns None if the config is unavailable or the entity
    is not found (e.g. deleted after the event was logged).
    """
    if entity_id is None:
        return None
    try:
        try:
            from .config_loader import get_config
        except ImportError:
            from config_loader import get_config
        config = get_config()
        for item in config.get(collection, []):
            if item.get("id") == entity_id:
                return item.get("name")
    except Exception:
        pass
    return None


def log_event(
    request_id: str,
    session_id: str,
    identity_id: Optional[int],
    team_id: Optional[int],
    server_id: Optional[int],
    jsonrpc_method: str,
    tool: Optional[str] = None,
    resource_uri: Optional[str] = None,
    prompt: Optional[str] = None,
    action: str = "allow",
    status: str = "ok",
    latency_ms: Optional[int] = None,
    error: Optional[str] = None,
    bytes_in: Optional[int] = None,
    bytes_out: Optional[int] = None,
    dlp_hits: Optional[list] = None,
    guardrail_hits: Optional[list] = None,
    params: Optional[dict] = None,
    result: Optional[Any] = None,
) -> None:
    """Write a single NDJSON event line to the events log.

    Identity, team, and server names are resolved from the config bundle so
    the NDJSON file contains human-readable labels alongside the numeric IDs.
    """
    path = _get_log_path()

    # Resolve names from the config bundle (best-effort; None if not found).
    identity_name = _lookup_name("identities", identity_id) if identity_id is not None else None
    team_name = _lookup_name("teams", team_id) if team_id is not None else None
    server_name = _lookup_name("servers", server_id) if server_id is not None else None

    event = {
        "ts": datetime.now(timezone.utc).isoformat(),
        "request_id": request_id,
        "session_id": session_id,
        "identity_id": identity_id,
        "identity_name": identity_name,
        "team_id": team_id,
        "team_name": team_name,
        "server_id": server_id,
        "server_name": server_name,
        "method": jsonrpc_method,
        "tool": tool,
        "resource_uri": resource_uri,
        "prompt": prompt,
        "action": action,
        "status": status,
        "latency_ms": latency_ms,
        "error": error,
        "bytes_in": bytes_in,
        "bytes_out": bytes_out,
        "dlp_hits": dlp_hits,
        "guardrail_hits": guardrail_hits,
    }

    if _is_log_payloads():
        if params is not None:
            try:
                event["params"] = json.dumps(params, default=str)[:4096]
            except Exception:
                event["params"] = None
        if result is not None:
            try:
                event["result"] = json.dumps(result, default=str)[:4096]
            except Exception:
                event["result"] = None

    try:
        os.makedirs(os.path.dirname(path), exist_ok=True)
        with open(path, "a", encoding="utf-8") as f:
            f.write(json.dumps(event, default=str) + "\n")
    except Exception as e:
        logger.warning("Failed to write MCP event log: %s", e)

File: mcp-gateway-rs/crates/gateway/src/events.rs
//! NDJSON event logger with 2-file rotation.
//!
//! Mirrors `mcp-gateway/events.py` but with 2-file rotation (not 10):
//! - Active file: `MCP_EVENTS_LOG_PATH` (default `data/mcp/events.ndjson`).
//! - Rotated copy: `{path}.1` (overwritten on rotation).
//! - When the active file exceeds `max_bytes`, it rotates to `.1`.
//! - One NDJSON line per RPC with method, tool, identity, team, server, etc.

use std::io::Write;
use std::path::PathBuf;
use std::sync::Arc;

use parking_lot::Mutex;
use serde::Serialize;

const DEFAULT_MAX_BYTES: u64 = 100 * 1024 * 1024; // 100 MB per file

/// Event record written as one NDJSON line.
#[derive(Debug, Serialize)]
pub struct Event {
    pub ts: String,
    pub request_id: String,
    pub session_id: String,
    pub identity_id: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub identity_name: Option<String>,
    pub team_id: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub team_name: Option<String>,
    pub server_id: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub server_name: Option<String>,
    pub method: String,
    pub tool: Option<String>,
    pub resource_uri: Option<String>,
    pub prompt: Option<String>,
    pub action: String,
    pub status: String,
    pub latency_ms: Option<u64>,
    pub error: Option<String>,
    pub bytes_in: Option<u64>,
    pub bytes_out: Option<u64>,
    pub dlp_hits: Option<serde_json::Value>,
    pub guardrail_hits: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<String>,
}

/// Thread-safe event logger with 2-file rotation.
#[derive(Clone)]
pub struct EventLogger {
    inner: Arc<Mutex<EventLoggerInner>>,
    log_payloads: bool,
}

struct EventLoggerInner {
    path: PathBuf,
    rotated_path: PathBuf,
    max_bytes: u64,
    current_size: u64,
}

impl EventLogger {
    pub fn new() -> Self {
        let path_str = std::env::var("MCP_EVENTS_LOG_PATH").unwrap_or_else(|_| "data/mcp/events.ndjson".into());
        let path = PathBuf::from(&path_str);
        let rotated_path = PathBuf::from(format!("{path_str}.1"));
        let max_bytes = std::env::var("MCP_EVENTS_MAX_BYTES")
            .ok()
            .and_then(|v| v.parse().ok())
            .unwrap_or(DEFAULT_MAX_BYTES);
        let log_payloads = std::env::var("MCP_LOG_PAYLOADS")
            .map(|v| matches!(v.to_lowercase().as_str(), "true" | "1" | "yes"))
            .unwrap_or(false);
        let current_size = std::fs::metadata(&path).map(|m| m.len()).unwrap_or(0);
        Self {
            inner: Arc::new(Mutex::new(EventLoggerInner {
                path,
                rotated_path,
                max_bytes,
                current_size,
            })),
            log_payloads,
        }
    }

    /// Generate a unique request ID (16 hex chars).
    pub fn generate_request_id() -> String {
        use uuid::Uuid;
        Uuid::new_v4().simple().to_string()[..16].to_string()
    }

    /// Write a single NDJSON event line.
    pub fn log(&self, mut event: Event) {
        // Strip params/result if not logging payloads.
        if !self.log_payloads {
            event.params = None;
            event.result = None;
        }
        let line = match serde_json::to_string(&event) {
            Ok(s) => s + "\n",
            Err(e) => {
                tracing::warn!("Failed to serialize event: {e}");
                return;
            }
        };
        let line_bytes = line.len() as u64;
        let mut inner = self.inner.lock();
        // Check rotation.
        if inner.current_size + line_bytes > inner.max_bytes {
            // Rotate: move current → rotated (overwrite).
            let _ = std::fs::rename(&inner.path, &inner.rotated_path)
                .or_else(|_| {
                    // Cross-device rename fails; fall back to copy + truncate.
                    std::fs::copy(&inner.path, &inner.rotated_path)
                        .and_then(|_| std::fs::write(&inner.path, ""))
                });
            inner.current_size = 0;
        }
        // Ensure parent dir exists.
        if let Some(parent) = inner.path.parent() {
            let _ = std::fs::create_dir_all(parent);
        }
        // Append.
        match std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&inner.path)
        {
            Ok(mut f) => {
                if let Err(e) = f.write_all(line.as_bytes()) {
                    tracing::warn!("Failed to write MCP event log: {e}");
                }
            }
            Err(e) => tracing::warn!("Failed to open MCP event log: {e}"),
        }
        inner.current_size += line_bytes;
    }
}

impl Default for EventLogger {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn request_id_is_16_hex() {
        let id = EventLogger::generate_request_id();
        assert_eq!(id.len(), 16);
        assert!(id.chars().all(|c| c.is_ascii_hexdigit()));
    }

    #[test]
    fn event_serializes_to_ndjson() {
        let e = Event {
            ts: "2025-01-01T00:00:00Z".into(),
            request_id: "abc123".into(),
            session_id: "sess1".into(),
            identity_id: Some(1),
            identity_name: Some("ci-bot".into()),
            team_id: Some(1),
            team_name: Some("Engineering".into()),
            server_id: Some(1),
            server_name: Some("jira".into()),
            method: "tools/call".into(),
            tool: Some("jira__create".into()),
            resource_uri: None,
            prompt: None,
            action: "allow".into(),
            status: "ok".into(),
            latency_ms: Some(42),
            error: None,
            bytes_in: None,
            bytes_out: Some(1024),
            dlp_hits: None,
            guardrail_hits: None,
            params: None,
            result: None,
        };
        let s = serde_json::to_string(&e).unwrap();
        assert!(s.contains("\"method\":\"tools/call\""));
        assert!(s.contains("\"tool\":\"jira__create\""));
        assert!(s.contains("\"identity_name\":\"ci-bot\""));
        assert!(s.contains("\"team_name\":\"Engineering\""));
        assert!(s.contains("\"server_name\":\"jira\""));
    }
}
